src/wire/igmpv2.rs

```diff
@@ -0,0 +1,451 @@
+
+use {Error, Result};
+use super::ip::checksum;
+use time::Duration;
+
+use wire::Ipv4Address;
+
+enum_with_unknown! {
+    /// Internet Group Management Protocol v2 message version/type.
+    pub doc enum Message(u8) {
+        /// Membership Query
+        MembershipQuery = 0x11,
+        /// Version 2 Membership Report
+        MembershipReportV2 = 0x16,
+        /// Leave Group
+        LeaveGroup = 0x17,
+        /// Version 1 Membership Report
+        MembershipReportV1 = 0x12
+    }
+}
+
+/// A read/write wrapper around an Internet Group Management Protocol v2 packet buffer.
+#[derive(Debug)]
+pub struct Packet<T: AsRef<[u8]>> {
+    buffer: T,
+}
+
+mod field {
+    use wire::field::*;
+
+    pub const TYPE: usize = 0;
+    pub const MAX_RESP_TIME: usize = 1;
+    pub const CHECKSUM: Field = 2..4;
+    pub const GROUP_ADDRESS: Field = 4..8;
+}
+
+impl fmt::Display for Message {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &Message::MembershipQuery => write!(f, "membership query"),
+            &Message::MembershipReportV2 => write!(f, "version 2 membership report "),
+            &Message::LeaveGroup => write!(f, "leave group"),
+            &Message::MembershipReportV1 => write!(f, "version 1 membership report"),
+            &Message::Unknown(id) => write!(f, "Unknown message {}", id),
+        }
+    }
+}
+
+
+/// Internet Group Management Protocol v2
+/// defined in [RFC 2236]
+///
+/// [RFC 2236]: https://tools.ietf.org/html/rfc2236
+impl<T: AsRef<[u8]>> Packet<T> {
+    /// Imbue a raw octet buffer with IGMPv2 packet structure.
+    pub fn new(buffer: T) -> Packet<T> {
+        Packet { buffer }
+    }
+
+    /// Shorthand for a combination of [new] and [check_len].
+    ///
+    /// [new]: #method.new
+    /// [check_len]: #method.check_len
+    pub fn new_checked(buffer: T) -> Result<Packet<T>> {
+        let packet = Self::new(buffer);
+        packet.check_len()?;
+        Ok(packet)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    /// Returns `Err(Error::Truncated)` if the buffer is too short.
+    pub fn check_len(&self) -> Result<()> {
+        let len = self.buffer.as_ref().len();
+        if len < field::GROUP_ADDRESS.end as usize {
+            Err(Error::Truncated)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Consume the packet, returning the underlying buffer.
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the message type field.
+    #[inline]
+    pub fn msg_type(&self) -> Message {
+        let data = self.buffer.as_ref();
+        Message::from(data[field::TYPE])
+    }
+
+    /// Return the Max reponse time (in 100ms units)
+    ///
+    /// See [RFC 3376]: 4.1.1. Max Resp Code
+    ///
+    /// [RFC 3376]: https://tools.ietf.org/html/rfc3376
+    #[inline]
+    pub fn max_resp_time(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::MAX_RESP_TIME].into()
```

Comment:
I don't think you need the `into` here if you're returning a `u8`. Did you mean to return a `Duration`?

Reply:
Right, thanks for pointing out!